File: app/src/main/java/com/marv42/ebt/newnote/CommentSuggestion.java
/*
 Copyright (c) 2010 - 2020 Marvin Horter.
 All rights reserved. This program and the accompanying materials
 are made available under the terms of the GNU Public License v2.0
 which accompanies this distribution, and is available at
 http://www.gnu.org/licenses/old-licenses/gpl-2.0.html
 */

package com.marv42.ebt.newnote;

import android.content.Context;
import android.os.AsyncTask;
import android.widget.Toast;

import androidx.core.util.Pair;

import com.marv42.ebt.newnote.data.LocationValues;
import com.marv42.ebt.newnote.data.LoginInfo;
import com.marv42.ebt.newnote.exceptions.CallResponseException;
import com.marv42.ebt.newnote.exceptions.HttpCallException;

import org.jetbrains.annotations.NotNull;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashSet;
import java.util.List;

import static android.widget.Toast.LENGTH_LONG;
import static com.marv42.ebt.newnote.exceptions.ErrorMessage.ERROR;

class CommentSuggestion extends AsyncTask<LocationValues, Void, String[]> {

    private static final String DATA_ELEMENT = "data";
    private static final String AMOUNT_ELEMENT = "amount";
    private static final String COMMENT_ELEMENT = "comment";
    private ApiCaller apiCaller;
    private Callback callback;
    private String additionalComment;

    CommentSuggestion(ApiCaller apiCaller, Callback callback, EncryptedPreferenceDataStore dataStore) {
        this.apiCaller = apiCaller;
        this.callback = callback;
        additionalComment = dataStore.get(R.string.pref_settings_comment_key, "")
                .replace("\u00a0", " ");
    }

    @Override
    protected String[] doInBackground(LocationValues... locationValues) {
        try {
            return getSuggestions(locationValues);
        } catch (CallResponseException e) {
            return getErrorStrings(e.getMessage());
        }
    }

    @NotNull
    private String[] getSuggestions(LocationValues[] locationValues) throws CallResponseException {
        JSONObject json = getJson(locationValues);
        JSONArray allSuggestions = json.optJSONArray(DATA_ELEMENT);
        List<String> uniques = getUniquesWrtAdditionalComments(allSuggestions);
        uniques = new ArrayList<>(new LinkedHashSet<>(uniques));
        return uniques.toArray(new String[0]);
    }

    private JSONObject getJson(LocationValues[] locationValues) throws CallResponseException {
        try {
            LoginInfo loginInfo = apiCaller.callLogin();
            List<Pair<String, String>> params = getLocationParams(loginInfo.sessionId, locationValues);
            String body = apiCaller.callMyComments(params);
            JSONObject json = new JSONObject(body);
            if (!json.has(DATA_ELEMENT))
                throw new CallResponseException("R.string.server_error: no '" + DATA_ELEMENT + "' element");
            return json;
        } catch (HttpCallException | CallResponseException | JSONException e) {
            throw new CallResponseException(e.getMessage());
        }
    }

    @NotNull
    private List<String> getUniquesWrtAdditionalComments(JSONArray allSuggestions) {
        List<JSONObject> suggestions = getJsonList(allSuggestions);
        Collections.sort(suggestions, (j1, j2) -> j2.optInt(AMOUNT_ELEMENT) - j1.optInt(AMOUNT_ELEMENT));
        List<String> uniques = new ArrayList<>();
        for (int i = 0; i < suggestions.size(); ++i) {
            String value = suggestions.get(i).optString(COMMENT_ELEMENT).replace("\u00a0", " ");
            if (value.endsWith(additionalComment))
                value = value.substring(0, value.length() - additionalComment.length());
            if (value.length() > 0 && !uniques.contains(value))
                uniques.add(value);
        }
        return uniques;
    }

    @NotNull
    private List<JSONObject> getJsonList(JSONArray allComments) {
        List<JSONObject> list = new ArrayList<>();
        for (int i = 0; allComments != null && i < allComments.length(); ++i)
            list.add(allComments.optJSONObject(i));
        return list;
    }

    @NotNull
    private String[] getErrorStrings(String s) {
        return new String[]{ERROR, s};
    }

    @NotNull
    private List<Pair<String, String>> getLocationParams(String sessionId, LocationValues[] locationValues) {
        LocationValues lv = locationValues[0];
        List<Pair<String, String>> params = new ArrayList<>();
        params.add(new Pair<>("m", "mycomments"));
        params.add(new Pair<>("v", "1"));
        params.add(new Pair<>("PHPSESSID", sessionId));
        params.add(new Pair<>("country", lv.country));
        params.add(new Pair<>("city", lv.city));
        params.add(new Pair<>("zip", lv.postalCode));
        return params;
    }

    @Override
    protected void onPostExecute(String[] s) {
        Context context = (Context) callback;
        if (s == null || s.length < 1) {
            Toast.makeText(context, context.getString(R.string.no_comment_suggestions), LENGTH_LONG).show();
            return;
        }
        if (s[0].equals(ERROR)) {
            Toast.makeText(context, context.getString(R.string.comment_suggestions_error), LENGTH_LONG).show();
            return;
        }
        Toast.makeText(context, context.getString(R.string.comment_suggestions_set), LENGTH_LONG).show();
        callback.onSuggestions(s);
    }

    interface Callback {
        void onSuggestions(String[] suggestions);
    }
}

File: app/src/main/java/com/marv42/ebt/newnote/CommentSuggestion.kt
/*
 Copyright (c) 2010 - 2022 Marvin Horter.
 All rights reserved. This program and the accompanying materials
 are made available under the terms of the GNU Public License v2.0
 which accompanies this distribution, and is available at
 http://www.gnu.org/licenses/old-licenses/gpl-2.0.html
 */
package com.marv42.ebt.newnote

import android.content.Context
import android.widget.Toast
import android.widget.Toast.LENGTH_LONG
import androidx.core.util.Pair
import com.marv42.ebt.newnote.data.LocationValues
import com.marv42.ebt.newnote.exceptions.CallResponseException
import com.marv42.ebt.newnote.exceptions.ErrorMessage
import com.marv42.ebt.newnote.exceptions.HttpCallException
import com.marv42.ebt.newnote.preferences.EncryptedPreferenceDataStore
import kotlinx.coroutines.MainScope
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.util.*

internal class CommentSuggestion(private val apiCaller: ApiCaller, private val callback: Callback, dataStore: EncryptedPreferenceDataStore) {

    private val additionalComment: String = dataStore.get(R.string.pref_settings_comment_key, "")
            .replace("\u00a0", " ")
    private val scope = MainScope()

    fun execute(locationValues: LocationValues) {
        scope.executeAsyncTask(
           onPreExecute = {
        }, doInBackground = {
            doInBackground(locationValues)
        }, onPostExecute = {
            onPostExecute(it)
        })
    }

    private fun doInBackground(locationValues: LocationValues): Array<String?>? {
        return try {
            getSuggestions(locationValues)
        } catch (e: CallResponseException) {
            getErrorStrings(e.message)
        }
    }

    @Throws(CallResponseException::class)
    private fun getSuggestions(locationValues: LocationValues): Array<String?> {
        val json = getJson(locationValues)
        val allSuggestions = json.optJSONArray(DATA_ELEMENT)
        var uniques = getUniquesWrtAdditionalComments(allSuggestions)
        uniques = ArrayList(LinkedHashSet(uniques))
        return uniques.toTypedArray()
    }

    @Throws(CallResponseException::class)
    private fun getJson(locationValues: LocationValues): JSONObject {
        return try {
            val loginInfo = apiCaller.callLogin()
            val params = getLocationParams(loginInfo.sessionId, locationValues)
            val body = apiCaller.callMyComments(params)
            val json = JSONObject(body)
            if (!json.has(DATA_ELEMENT))
                throw CallResponseException("R.string.server_error: no '$DATA_ELEMENT' element")
            json
        } catch (e: HttpCallException) {
            throw CallResponseException(e.message)
        } catch (e: CallResponseException) {
            throw CallResponseException(e.message)
        } catch (e: JSONException) {
            throw CallResponseException(e.message)
        }
    }

    private fun getUniquesWrtAdditionalComments(allSuggestions: JSONArray?): List<String?> {
        val suggestions = getJsonList(allSuggestions)
        suggestions.sortWith { j1: JSONObject, j2: JSONObject -> j2.optInt(AMOUNT_ELEMENT) - j1.optInt(AMOUNT_ELEMENT) }
        val uniques: MutableList<String?> = ArrayList()
        for (i in suggestions.indices) {
            var value = suggestions[i].optString(COMMENT_ELEMENT).replace("\u00a0", " ")
            if (value.endsWith(additionalComment))
                value = value.substring(0, value.length - additionalComment.length)
            if (value.isNotEmpty() && !uniques.contains(value))
                uniques.add(value)
        }
        return uniques
    }

    private fun getJsonList(allComments: JSONArray?): MutableList<JSONObject> {
        val list: MutableList<JSONObject> = ArrayList()
        var i = 0
        while (allComments != null && i < allComments.length()) {
            list.add(allComments.optJSONObject(i))
            ++i
        }
        return list
    }

    private fun getErrorStrings(s: String?): Array<String?> {
        return arrayOf(ErrorMessage.ERROR, s)
    }

    private fun getLocationParams(sessionId: String, locationValues: LocationValues): List<Pair<String, String>> {
        val params: MutableList<Pair<String, String>> = ArrayList()
        params.add(Pair("m", "mycomments"))
        params.add(Pair("v", "1"))
        params.add(Pair("PHPSESSID", sessionId))
        params.add(Pair("country", locationValues.country))
        params.add(Pair("city", locationValues.city))
        params.add(Pair("zip", locationValues.postalCode))
        return params
    }

    private fun onPostExecute(s: Array<String?>?) {
        val context = callback as Context
        if (s == null || s.isEmpty()) {
            Toast.makeText(context, R.string.no_comment_suggestions, LENGTH_LONG).show()
            return
        }
        if (s[0] == ErrorMessage.ERROR) {
            Toast.makeText(context, R.string.comment_suggestions_error, LENGTH_LONG).show()
            return
        }
        Toast.makeText(context, R.string.comment_suggestions_set, LENGTH_LONG).show()
        callback.onSuggestions(s)
    }

    internal interface Callback {
        fun onSuggestions(suggestions: Array<String?>?)
    }

    companion object {
        private const val DATA_ELEMENT = "data"
        private const val AMOUNT_ELEMENT = "amount"
        private const val COMMENT_ELEMENT = "comment"
    }
}